### PR TITLE
Adds `as_completed` utility for `PrefectFuture`

### LINKED ISSUE #14241: Add `as_completed` utility for `PrefectFuture`s
Add a wait utility that mirrors concurrent.futures.as_completed. Should accept an iterable of futures to wait for completion and a timeout. It should act as a generator and yield futures as they are completed.

### PR DESCRIPTION
**Pull request by jeanluciano**
Adds a utility that yields completed futures in order they are completed. Closes #14241 

### Example
```python
        @task
        def sleep_task(seconds):
            sleep(seconds)
            return 42

        @flow
        def flow():
            futures = random_task.map(range(10))
            for future in as_completed(futures):
                print(future.result())
```

### Checklist

- [ x ] This pull request includes a label categorizing the change e.g. `maintenance`, `fix`, `feature`, `enhancement`, `docs`.
- [ x ] This pull request references any related issue by including "closes `<link to issue>`"
  
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [ x ] If this pull request adds new functionality, it includes unit tests that cover the changes
- [ x ] If this pull request removes docs files, it includes redirect settings in `mint.json`.
- [ x ] If this pull request adds functions or classes, it includes helpful docstrings.

### CODE REVIEW

File: src/prefect/futures.py [hunk 1]
```diff
@@ -292,6 +292,59 @@ def result(
             ) from exc
 
 
+def as_completed(futures: List[PrefectFuture], timeout=None) -> Iterator[PrefectFuture]:
+    """
+    An iterator over the given futures that yields each as it completes.
+
+    Args:
+        fs: The sequence of Futures (possibly created by different Executors) to
```
Comment (zzstoatzz) [inline src/prefect/futures.py:300]:
```suggestion
        futures: The sequence of Futures (possibly created by different Executors) to
```

File: src/prefect/futures.py [hunk 2]
```diff
@@ -292,6 +292,59 @@ def result(
             ) from exc
 
 
+def as_completed(futures: List[PrefectFuture], timeout=None) -> Iterator[PrefectFuture]:
```
Comment (zzstoatzz) [inline src/prefect/futures.py:295]: i'd double check my suggestion, but this looks like its a Generator function right?
```suggestion
def as_completed(futures: List[PrefectFuture], timeout: Optional[float] = None) -> Generator[PrefectFuture, None]:
```

File: src/prefect/futures.py [hunk 3]
```diff
@@ -292,6 +292,59 @@ def result(
...
+            return 42
+
+        @flow
+        def flow():
+            futures = random_task.map(range(10))
+            for future in as_completed(futures):
+                print(future.result())
+        ```
+    """
+    futures = set(futures)
```
Comment (zzstoatzz) [inline src/prefect/futures.py:328]: this overshadows the `futures` input arg, can we call this `unique_futures` or something?

**Comment by desertaxle** [comment]
I think this implementation will be good for now, but it looks like futures early in the list will block futures later in the list from being yielded when they are complete until the futures earlier in the list are finished. Could you see if an implementation that would allow the behavior of this utility to not be dependent on the ordering of the passed in futures? I think a callback structure on `PrefectFuture` might make that possible.

File: src/prefect/futures.py [hunk 4]
```diff
@@ -150,6 +163,17 @@ def result(
             _result = run_coro_as_sync(_result)
         return _result
 
+    def add_done_callback(self, fn: Callable[[PrefectFuture], None]):
+        if not self._final_state:
+
+            def call_with_self(future):
+                logger.debug(f"{future.result()}")
```
Comment (cicdw) [inline src/prefect/futures.py:170]:
```suggestion
```

File: src/prefect/futures.py [hunk 5]
```diff
@@ -292,6 +339,47 @@ def result(
...
+
+            while pending:
+                event.wait()
+                with lock:
+                    done = finished_futures
+                    finished_futures = []
+                    event.clear()
+
+                for future in done:
+                    logger.debug(f"done: {future.result()}")
```
Comment (cicdw) [inline src/prefect/futures.py:373]:
```suggestion
```

File: src/prefect/task_runs.py [hunk 6]
```diff
@@ -195,6 +195,48 @@ async def wait_for_task_run(
...
+
+        try:
+            # Now check one more time whether the task run arrived before we start to
+            # wait on it, in case it came in while we were setting up the event above.
+            with instance._observed_completed_task_runs_lock:
+                if task_run_id in instance._observed_completed_task_runs:
+                    callback()
+                    return
+
+            await from_async.wait_for_call_in_loop_thread(
```
Comment (cicdw) [inline src/prefect/task_runs.py:231]: we're trying to move away from using these utilities - is this one necessary?

**Review by cicdw — CHANGES_REQUESTED** [review]
a few stray debug logs that I think need to be removed, and a question

**Comment by desertaxle** [comment]
FYI, I think we'll want to support this for the Dask and Ray flavors of `PrefectFuture` also

File: src/prefect/futures.py [hunk 7]
```diff
@@ -91,6 +92,16 @@ def result(
...
 
+    @abc.abstractmethod
+    def add_done_callback(self, fn):
+        ...
+        """
+            Add a callback to be run when the future completes or is cancelled.
+
+            Args:
+                fn: A callable that will be called with this future as its only argument when the future completes or is cancelled.
+            """
```
Comment (zzstoatzz) [inline src/prefect/futures.py:103]: nit: I think the ellipsis should be after the docstring here

File: src/prefect/futures.py [hunk 8]
```diff
@@ -150,6 +162,16 @@ def result(
             _result = run_coro_as_sync(_result)
         return _result
 
+    def add_done_callback(self, fn: Callable[[PrefectFuture], None]):
```
Comment (zzstoatzz) [inline src/prefect/futures.py:165]: maybe we could have a docstring here to briefly explain the `call_with_self` business here?

File: src/prefect/futures.py [hunk 9]
```diff
@@ -292,6 +333,46 @@ def result(
             ) from exc
 
 
+def as_completed(
+    futures: List[PrefectFuture], timeout: Optional[float] = None
+) -> Generator[PrefectFuture, None]:
```
Comment (zzstoatzz) [inline src/prefect/futures.py:338]: seems pretty elegant to me!

**Review by zzstoatzz — COMMENTED** [review]
nice!

a couple nits and:

maybe we should add a test or two for cases where we throw exceptions / timeout? just to be explicit on what we expect

File: src/integrations/prefect-dask/tests/test_task_runners.py [hunk 10]
```diff
@@ -248,6 +249,25 @@ def test_flow():
...
+            return seconds
+
+        timings = [1, 5, 10]
+        with task_runner:
+            done_futures = []
+            futures = [
+                task_runner.submit(
+                    sleep_task, parameters={"seconds": seconds}, wait_for=[]
+                )
+                for seconds in timings
```
Comment (cicdw) [inline src/integrations/prefect-dask/tests/test_task_runners.py:265]: I think this test would be more compelling if the timings were ordered in reverse, proving that `as_completed` actually affected the order that they yielded.
```suggestion
                for seconds in reversed(timings)
```

File: src/integrations/prefect-ray/tests/test_task_runners.py [hunk 11]
```diff
@@ -371,6 +372,24 @@ def test_flow():
...
+        def task_a(seconds):
+            time.sleep(seconds)
+            return seconds
+
+        timings = [1, 5, 10]
+
+        @flow(version="test", task_runner=task_runner)
+        def test_flow():
+            done_futures = []
+            futures = [task_a.submit(seconds) for seconds in timings]
```
Comment (cicdw) [inline src/integrations/prefect-ray/tests/test_task_runners.py:386]: same comment as above
```suggestion
            futures = [task_a.submit(seconds) for seconds in reversed(timings)]
```

File: src/integrations/prefect-dask/prefect_dask/task_runners.py [hunk 12]
```diff
@@ -149,6 +149,16 @@ def result(
             _result = run_coro_as_sync(_result)
         return _result
 
+    def add_done_callback(self, fn: Callable[[PrefectFuture], None]):
```
Comment (cicdw) [inline src/integrations/prefect-dask/prefect_dask/task_runners.py:152]: I don't think it's necessary to add these methods for each task runner given that `PrefectWrappedFuture` already has it implemented

File: src/prefect/task_runs.py [hunk 13]
```diff
@@ -195,6 +198,26 @@ async def wait_for_task_run(
...
+        instance = cls.instance()
+        with instance._observed_completed_task_runs_lock:
+            if task_run_id in instance._observed_completed_task_runs:
+                callback()
+                return
+
+        with instance._completion_events_lock:
+            # Cache the event for the task run ID so the consumer can set it
+            # when the event is received
+            instance._completion_events[task_run_id] = callback
```
Comment (cicdw) [inline src/prefect/task_runs.py:219]: this doesn't fit the typing on `instance._completion_events` - is this intentional?

File: tests/test_futures.py [hunk 14]
```diff
@@ -37,6 +39,9 @@ def result(
     ) -> Any:
         return self._final_state.result()
 
+    def add_done_callback(self, fn):
+        return super().add_done_callback(fn)
```
Comment (cicdw) [inline tests/test_futures.py:43]: is this necessary or can you simply inherit this method?

File: tests/test_futures.py [hunk 15]
```diff
@@ -55,6 +60,79 @@ def test_wait_with_timeout(self):
...
+            import time
+
+            time.sleep(seconds)
+            return seconds
+
+        with ThreadPoolTaskRunner() as runner:
+            futures = []
+            timings = [1, 5, 10]
+
+            for i in timings:
```
Comment (cicdw) [inline tests/test_futures.py:96]:
```suggestion
            for i in reversed(timings):
```

File: tests/test_futures.py [hunk 16]
```diff
@@ -55,6 +60,79 @@ def test_wait_with_timeout(self):
...
+        @task
+        async def my_task(seconds):
+            import time
+
+            time.sleep(seconds)
+            return seconds
+
+        futures = []
+        timings = [1, 5, 10]
+        for i in timings:
```
Comment (cicdw) [inline tests/test_futures.py:116]:
```suggestion
        for i in reversed(timings):
```

File: tests/test_task_runners.py [hunk 17]
```diff
@@ -61,6 +61,9 @@ def result(
         self.wait()
         return self._state.result()
 
+    def add_done_callback(self, fn):
```
Comment (cicdw) [inline tests/test_task_runners.py:64]: necessary?

**Review by cicdw — CHANGES_REQUESTED** [review]
some small nitpicks about tests and duplicated code, the core implementation makes sense to me

**Comment by cicdw** [comment]
Really weird that Ray is behaving strangely; @jeanluciano I don't think we need to optimize for a perfectly reversed output, as there is probably some latency in spinup that is allowing the 5 second sleep to execute faster.  Maybe just assert that `10` is the last one returned, which will prove that `as_completed` affected the return order (ultimately all we need for this specific test).

**Comment by cicdw** [comment]
I would keep the Dask test as-is and only change the Ray test, between the two it's decent coverage

**Comment by jeanluciano** [comment]
Yeah I thought it had to do with the way I was converting the object ref into a future, but setting the callback on the object ref directly behaves the same way.

**Review by cicdw — APPROVED** [review]

**Pull request merged**